integrations: Add `interfaced_setting` framework.

PieterCK opened the pull request:
Overview
---
This adds **Interfaced settings**, those are pre-built API endpoint parameter types that come with a custom UI in the frontend. To use an interfaced setting, an integration would just have to add queries typed with the interfaced setting (for now its only `MapToChannelsT`).

### The `MapToChannelsT` interfaced setting.
The UI that comes with `MapToChannelT`  is the "Map to Zulip Channels" option in the "Where to send notifications" field in the Generate integration URL modal. This is planned to replace Slack integrations `channels_map_to_topics` setting, but it is applicable for other third-party chat integrations too.

**How to use it:**
Use the supported interfaced settings in the endpoint query type.
```python

from zerver.lib.webhooks.interfaced_settings import MapToChannelsT
...

@webhook_view("Slack", notify_bot_owner_on_invalid_json=False)
@typed_endpoint
def api_slack_webhook(
    request: HttpRequest,
    user_profile: UserProfile,
    *,
    slack_app_token: str = "",
    channels_mapping: MapToChannelT = False, # <-- This here!

```
**Result:**

| Lightbox | Dark |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/bd68a442-e0a4-4a3f-a49f-ba68b40b9be4) | ![image](https://github.com/user-attachments/assets/84fc48d5-2c87-4dd4-a4fa-0389105cf071) | 

**Just to prove the point that any integration could use this setting (where applicable):**

| Example 1 | Example 2 |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/c6a99d19-fd75-4898-b29e-0ee42d7b617f) | ![image](https://github.com/user-attachments/assets/609b255a-443b-4495-a91a-8dc0f94dec91) |

**Behaviours**
If selected:
- The  "Send all notifications to a single topic" field is disabled
- The setting's specific query is also added to the integration URL
Video of manual test:

[Screencast from 2025-02-12 13-24-32.webm](https://github.com/user-attachments/assets/eb85fafc-00ec-417d-b1ec-d69b0db0be2d)

---
| API doc update |
|--------|
| ![image](https://github.com/user-attachments/assets/b316ed00-5e8a-482c-8f31-8d4008b43300) `/api/register-queue`| 

Refactor on `config_options` and `all_event_types`
---
This PR also refactors some frontend logic and `state_data.ts` schema for `config_options` and `all_event_types`.  

<details>

<summary>
Here are some screenshots as proof that there are no regressions to these features:  
</summary>

| feature | screenshot |
|--------|--------|
| `config_options`  | ![image](https://github.com/user-attachments/assets/577aa167-f9a3-4815-9f63-a7d4ecabb8ca)|
| `all_event_types` | ![image](https://github.com/user-attachments/assets/2f02f60d-3242-47a4-a348-ea5c1f719ee0) |

</details>

Concerns
---
- The `integration_url_modal.ts` doesn't have any node test file, I wonder what the priority of adding that is. Having to support 3 or so different frameworks (`all_event_types`, `config_options` & `interfaced_settings`) on top of its main functions I think it'd be great if we could have a test suite for it.

Issue discussion that leads to this solution: [CZO](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Slack's.20channels_map_to_topics.20setting.2E/with/2072236)
Design discussion: CZO


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
Yep. Most of the strings and UI details can be assumed as placeholders for the moment. I'll post an update and bring the prototype to #integration or #design to be iterated--for both design and technical approach--once it's a bit more presentable : )

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -61,6 +61,10 @@ export function show_generate_integration_url_modal(api_key: string): void {
     };
+    const map_to_channels_option = {
+        name: $t_html({defaultMessage: "Map to Zulip channels"}),
+        unique_id: -2,
```
`unique_id: -2` feels hacky to me... I wonder if we need a more proper structure for non-stream options or if this is enough.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -183,7 +209,12 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            const interfaced_settings = selected_integration_data?.interfaced_settings;
+            const map_to_channel_setting = interfaced_settings?.MapToChannelsT;
+
+            if (map_to_channel_setting && stream_id === -2) {
```
`map_to_channels_option` having `unique_id: -2` feels shoddy to me... It's especially apparent here. I wonder if we need a more proper structure for non-stream options or if this is enough.

timabbott replied:
Yeah, probably the natural refactor there would be for `unique_id` to have a type that is a `int | str`, with the strings used for the special values. We use that pattern for the channel input in the `compose_recipient` module, I think?

PieterCK replied:
ohh ok it is `string | number`:
```javascript
// dropdown_widget.ts

export type Option = {
    unique_id: number | string;
    name: string;
    is_setting_disabled?: boolean;
    stream?: StreamSubscription;
};
```

PieterCK replied:
So,  `DropdownWidget` expects its caller to explicitly declare its `unique_id`'s type in `unique_id_type`. Right now it accepts:
```javascript
export enum DataTypes {
    NUMBER = "number",
    STRING = "string",
}
```

`unique_id_type` is only used in this logic in `class DropdownWidget`:
```javascript
                // Click on item.
                $popper.on("click", ".list-item", (event) => {
                    const selected_unique_id = $(event.currentTarget).attr("data-unique-id");
                    assert(selected_unique_id !== undefined);
                    this.current_value = selected_unique_id;
                    if (this.unique_id_type === DataTypes.NUMBER) {
                        this.current_value = Number.parseInt(this.current_value, 10);
                    }
                    this.item_click_callback(event, instance, this, false);
                });
```
I wonder if it's better to refactor the on-click logic to use something like `typeof`? Are there any other reasons to explicitly declare what type of `unique_id` other than this logic?

PieterCK commented:
@zulipbot add "area: integrations"

Aditya8840 commented:
Manually tested the config options changes; looks good to me.

timabbott commented on zerver/lib/webhooks/interfaced_settings.py:
```diff
@@ -0,0 +1,75 @@
+    # other query the integration might declare.
+    param_id = NEXT_ID("param")
+    key = f"{param}_{param_id}"
+    return key
```
Can you explain how this algorithm works?

PieterCK replied:
Since the query name for these Types is tucked away behind a TypeAlias and the type itself is declared in another file, adding a code or ID to the query variable here is just to make sure we avoid accidentally declaring duplicate parameter names when using it in the integration endpoint.

But, now that I think about it again, probably a simple prefix format is enough, something unusual like `Z_{param}` e.g, `Z_map_to_channels`.

timabbott commented on zerver/tests/test_integrations.py:
```diff
@@ -83,3 +94,87 @@ def test_no_missing_screenshot_path(self) -> None:
+    )
+    @patch.object(mocked_integration, "get_function")
+    def test_unknown_setting(self, mock_endpoint: MagicMock) -> None:
+        def mock_api_unkown_setting_webhook(
```
`unknown` has a typo.

timabbott commented on zerver/lib/webhooks/interfaced_settings.py:
```diff
@@ -0,0 +1,75 @@
+
+    if settings_context is not None:
+        return asdict(settings_context)
+    raise AssertionError(f"Please add context for this setting: {arg_name}")
```
`Please define a SettingContext for this setting: {arg_name}` would be a lot more clear.

timabbott commented on zerver/lib/webhooks/interfaced_settings.py:
```diff
@@ -0,0 +1,75 @@
+# Keep this in sync with `state_data.integrations_interfaced_settings_schema`
+# this dict is not a dataclass because Type aliases inside dataclass definitions
+# are not supported at runtime.
+SUPPORTED_INTERFACED_SETTINGS: dict[str, TypeAlias] = {"MapToChannelsT": MapToChannelsT}
```
Let's line-wrap this with each setting on its own line.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -22896,6 +22898,57 @@ components:
+
+            If the option is selected, the `&map_to_channels_1=true` query
+            will be added to the integrations URL. Where `map_to_channels_1`
+            is the unique query of this setting.
```
OK, I think this should be explained more clearly as a `{unique_query}=true` parameter will be added to the integrations URL?

PieterCK replied:
Done, the paragraph is changed to this:
```
            If the option is selected, the `{unique_query}=true` query will be
            added to the integrations URL. Where the value of `unique_query` is,
            as the name suggests, the unique query of each interfaced setting.
```

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,12 @@ format used by the Zulip server that they are interacting with.
+
+* [`POST /register`](/api/register-queue): Updated
+  `realm_incoming_webhook_bots` with a new `interfaced_settings` key,
+  it defines what interfaced settings each integration uses.
```
You can migrate to the new `unmerged.d` system to avoid renumbering work for these. See the "api design" channel.

timabbott commented on web/src/integration_url_modal.ts:
```diff
@@ -13,23 +13,12 @@ import * as dialog_widget from "./dialog_widget.ts";
-});
-
-const config_options_schema = z.array(config_option_schema);
+type ConfigOption = z.infer<typeof integration_config_options_schema>;
```
I think this is now a bad name, given the type is an array of options. Probably you want to still define the singular as a type in `state_data.ts`.

PieterCK replied:
Updated the schema to this:

```javascript
export const integration_config_options_schema = z.object({
    key: z.string(),
    label: z.string(),
    validator: z.string(),
});

            // ...
            config_options: z.array(integration_config_options_schema).optional(),
```

timabbott commented on web/src/integration_url_modal.ts:
```diff
@@ -75,8 +64,11 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            const validated_config = config;
+            if (!validated_config) {
+                return;
+            }
```
This doesn't do any validation now? I'm not sure what you're trying to do.

PieterCK replied:
Updated to this:

```diff
diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
index 82bbe7540e..4a827533e5 100644
--- a/web/src/integration_url_modal.ts
+++ b/web/src/integration_url_modal.ts
@@ -14,13 +14,13 @@ import * as dialog_widget from "./dialog_widget.ts";
 import * as dropdown_widget from "./dropdown_widget.ts";
 import type {DropdownWidget, Option} from "./dropdown_widget.ts";
 import {$t_html} from "./i18n.ts";
-import type {integration_config_options_schema} from "./state_data.ts";
+import type {integration_config_option_schema} from "./state_data.ts";
 import {realm} from "./state_data.ts";
 import * as stream_data from "./stream_data.ts";
 import {place_caret_at_end} from "./ui_util.ts";
 import * as util from "./util.ts";
 
-type ConfigOption = z.infer<typeof integration_config_options_schema>;
+type ConfigOption = z.infer<typeof integration_config_option_schema>;
 
 export function show_generate_integration_url_modal(api_key: string): void {
     const default_url_message = $t_html({defaultMessage: "Integration URL will appear here."});
@@ -66,14 +66,10 @@ export function show_generate_integration_url_modal(api_key: string): void {
             );
         });
 
-        function render_config(config: ConfigOption): void {
-            const validated_config = config;
-            if (!validated_config) {
-                return;
-            }
+        function render_config(config: ConfigOption[]): void {
             $config_container.empty();
```

timabbott commented on web/src/state_data.ts:
```diff
@@ -255,6 +255,8 @@ export const realm_linkifier_schema = z.object({
     id: z.number(),
 });
 
+export const integration_all_event_types_shcmea = z.nullable(z.array(z.string()));
```
`schema` has a typo.

PieterCK replied:
I decided to drop this commit and declare `all_event_types` like this in `class IntegrationData`:

```javascript
type EventType = string;
...

class IntegrationData {
    seletec_integration: string;
    config_options: ConfigOption[] | undefined = undefined;
    all_event_types: EventType[] | null = null;
    interfaced_settings: InterfacedSetting | undefined = undefined;
```
the motivation for change is similar to the change for `type ConfigOption` in the previous commit, but I thought abstracting the schema into a variable like previously would just make it more obscure than clear.

timabbott commented on web/src/integration_url_modal.ts:
```diff
@@ -13,13 +13,42 @@ import * as dialog_widget from "./dialog_widget.ts";
+type InterfacedSetting = z.infer<typeof integrations_interfaced_settings_schema>;
+
+class IntegrationData {
+    seletec_integration: string;
```
`selected` has a typo.

timabbott commented:
Posted some feedback, I think this direction looks pretty plausible, though there's some things I don't understand.

PieterCK commented on zerver/openapi/zulip.yaml:
```diff
@@ -22896,6 +22898,57 @@ components:
+            unique_query:
+              type: string
+              description: |
+                The unique query variable of each interfaced setting.
```
Updated to this, I wonder if its clear enough:
```yaml
          properties:
            parameter_name:
              type: string
              description: |
                The parameter name declared in the integrations endpoint using
                this interfaced setting as its type.

                Clients can use this as additional context to enforce the custom
                behavior each setting promises.
            unique_query:
              type: string
              description: |
                The unique query variable of each interfaced setting. It is used
                as the query name when encoding the users input in the integration
                URL.
```

PieterCK commented:
Thanks for the feedback, I've posted how I addressed them in the individual review comments, and I've also updated the API doc screenshot in the PR description. I think I can start posting this somewhere in #design or #integration to get some input on the UI elements and strings?

PieterCK commented:
I opened a [#design](https://chat.zulip.org/#narrow/channel/101-design/topic/UI.20for.20integration's.20.22interfaced.20settings.22.20.2333370/with/2123070) discussion for the UI elements

PieterCK commented:
Closing this in favor or #34353
